File: rust/src/day1.rs
use crate::utils;

pub fn solution() {
    part_1()
}

fn part_1() {
    let lines = utils::lines("./resources/day1/input");
    let calories = calories_list(lines);
    println!("Day 1");
    println!("part1: {}", calories.first().unwrap());
    println!("part2: {}", calories.into_iter().take(3).sum::<i32>())
}

fn calories_list(lines: Vec<String>) -> Vec<i32> {
    let mut elves: Vec<i32> = Vec::new();
    let mut current_elf: i32 = 0;
    for line in lines {
        if line != "" {
            let calories = line.parse::<i32>().unwrap_or(0);
            current_elf += calories
        } else {
            elves.push(current_elf);
            current_elf = 0
        }
    }
    elves.push(current_elf);
    elves.sort_unstable();
    elves.reverse();
    return elves
}


#[cfg(test)]
mod tests {
    use crate::day1::calories_list;

    #[test]
    fn sums_elves_calories() {
        let input: Vec<String> = vec![str("1"), str("2"), str(""), str("2")];
        let result = calories_list(input);
        assert_eq!(result, vec![3, 2])
    }

    #[test]
    fn sorts_descending() {
        let input = vec![str("1"), str("2"), str(""), str("4"), str(""), str("3"), str("3")];
        let result = calories_list(input);
        assert_eq!(result, vec![6, 4, 3])
    }

    fn str(str: &str) -> String {
        String::from(str)
    }
}

File: 2022/clojure/src/aoc_2022/day1.clj
(ns aoc-2022.day1
  (:gen-class)
  (:require [aoc-2022.utils :as utils]))

(defn grouped-by-calories [lines]
  (filter #(not= '("") %) (partition-by #(= "" %) lines)))

(defn to-int [groups]
  (map #(map (fn [x] (Integer/parseInt x)) %) groups))

(defn calories-groups [lines]
  (to-int (grouped-by-calories lines)))

(defn sum-groups [grouped]
  (map #(reduce + 0 %) grouped))

(defn sort-desc [coll]
  (reverse (sort coll)))

(defn first-elf [elves]
  (first elves))

(defn first-three-elves-sum [elves]
  (reduce + (take 3 elves)))

(defn -main
  "Day 1"
  [& _]
  (let [elves (-> (utils/file-lines "./resources/day1/input") calories-groups sum-groups sort-desc)]
    (println (str "Highest elf: " (first-elf elves)))
    (println (str "first three: " (first-three-elves-sum elves)))))